website/blog/2022-08-22-unit-testing-dbt-package.md

```diff
@@ -0,0 +1,175 @@
+<!--truncate-->
+
+## Unit Testing vs. Integration Testing
+
+Many dbt packages use integration tests as their primary testing methodology. For example [dbt-utils](https://github.com/dbt-labs/dbt-utils) has [the integration_tests directory](https://github.com/dbt-labs/dbt-utils/tree/main/integration_tests) so that we can run integration tests by using the generic tests and macros contained within the package. The integration tests directory is essentially a standard dbt project within the dbt-utils package that is tested much the same way any dbt project would be.
+
+To use the integration tests - you’d simply run `dbt test` within the `integration_tests` directory. The tests execute as normal - meaning you can use your favorite methods of running CI against your dbt project to ensure that your integration tests are passing.
+
+Integration tests can help give you peace of mind that your package is performing as expected - but they have some drawbacks. Macros and generic tests frequently call other macros and the deeper dependency calls get, the more difficult it becomes to debug your macros using only integration tests.
+
+In this scenario it can be helpful to go beyond integration tests and implement unit tests for your macros. These unit tests can be run with a [dbt run operation](https://docs.getdbt.com/reference/commands/run-operation). Let’s take a look at a quick example of how this can be done.
+
+Consider a dbt package called dbt_sample_package . We would like to implement a simple macro to create a string literal from a string text in a macro named `to_literal` in the file `macros/to_literal.sql`.
+
+```sql
+- macros/to_literal.sql
+{% to_literal(text) %}
+'{{- text -}}'
+{% endmacro %}
+```
+
+---
+
+To implement a unit testing macro corresponding to the `to_literal` macro we can create a macro to test our original macro in `integration_tests/macros/test_to_literal.sql`.
+
+Then we call the `to_literal` macro in the testing macro. and if the result isn’t the same as expected, we raise an error using the [exceptions.raise_compiler_error macro](https://docs.getdbt.com/reference/dbt-jinja-functions/exceptions).
+
+```sql
+-- integration_tests/macros/test_to_literal.sql
+{% test_to_literal() %}
+	{% = dbt_sample_package.to_literal('test string') %}
+	{% if result != "'test string'" %}
+		{{ exceptions.raise_compiler_error('The test is failed') }}
+	{% endif %}
+{% endmacro %}
+```
+
+---
+
+By doing that, we can call the testing macro in the dbt project of integration tests using `dbt run-operation`.
+
+```sql
+dbt -operation test_to_literal
+```
+
+---
+
+If we want to run all tests with a single command, it would be good to bundle them in a macro. Moreover, we can call the macro with `dbt run-operation`.
+
+```sql
+-- integration_tests/macros/run_unit_tests.sql
+{% macro run_unit_tests() %}
+	{% do test_to_literal() %}
```

Comment:
```suggestion
	{% do test_to_literal() %}
	{% do another_test() %}
```

Or something like this to show that "bundling" power here

Reply:
ahh excellent idea